
# Assignment

In this assignment, we see how using convolutional neural networks improves our model's accuracy. Recall the fashion MNIST data set from the previous assignment:

We imported the libraries

In [1]:
import numpy as np
import matplotlib.pyplot as plt
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation, Flatten, Dropout

data = keras.datasets.fashion_mnist
(x_train, y_train), (x_test, y_test) = data.load_data()

Run the below cell to train a neural network that more or less imitates what a logistic regression classifier would do. 

In [2]:

model = Sequential()
model.add(Flatten(input_shape = (28, 28)))
model.add(Dense(10, activation = 'sigmoid'))

model.compile(
    optimizer = keras.optimizers.RMSprop(learning_rate = 1e-3),
    loss = keras.losses.SparseCategoricalCrossentropy(),
    metrics = [keras.metrics.SparseCategoricalAccuracy()],
)

model.fit(x_train, y_train, epochs = 20, verbose = 0)

model.evaluate(x_train, y_train, verbose = 2)
model.evaluate(x_test, y_test, verbose = 2)

2023-06-15 04:26:46.668462: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-06-15 04:26:46.741069: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2023-06-15 04:26:46.741759: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2200160000 Hz


1875/1875 - 1s - loss: 18.1423 - sparse_categorical_accuracy: 0.8075
313/313 - 0s - loss: 21.9740 - sparse_categorical_accuracy: 0.7871


[21.97396469116211, 0.7871000170707703]

### As we saw in the previous assignment, the above network seems to do a good job at classifying shirts, but not other clothing items. Let's see if we can improve the classifier by using convolutional layers.

- Copy the code in the above cell below, then modify the network so the first two layers are a convolution layer with 32 filters followed by a 5x5 max-pooling layer. Also add a dense layer with 32 neurons prior to the output layer. Train the new network for 50 epochs and report the accuracy. <span style="color:red" float:right>[5 point]</span>

we normalized our data and then we defined the model and add the layers.
First we reshape the input to be suitable for our modeling and then we add the layers based on the filters and the kernel and the channel is 1 becuase the image is in greyscale

In [3]:
x_train=x_train.astype('float32')/255.0
x_test=x_test.astype('float32')/255.0


In [4]:
from keras.layers.convolutional import Conv2D, MaxPooling2D
from keras.layers import Conv2D, Dense, Reshape


In [5]:
## your code goes here
model_1 = Sequential()
model_1.add(Reshape((28, 28, 1), input_shape=(28, 28)))

model_1.add(Conv2D(32, (3, 3), input_shape=(28, 28, 1), padding='same', activation='sigmoid'))
#model.add(Conv2D(32, (3, 3), activation='sigmoid', padding='valid'))
model_1.add(MaxPooling2D(pool_size=(5, 5)))
model_1.add(Flatten())
model_1.add(Dense(32, activation='relu'))
model_1.add(Dense(10, activation = 'sigmoid'))



In [6]:
#we definded the metrics.
model_1.compile(
    optimizer = keras.optimizers.RMSprop(learning_rate = 1e-3),
    loss = keras.losses.SparseCategoricalCrossentropy(),
    metrics = [keras.metrics.SparseCategoricalAccuracy()],
)

we trained the model and printed the results to see the numbers of the parameters.

In [7]:
model_1.fit(x_train, y_train, epochs = 50, verbose = 0)


In [8]:

model_1.evaluate(x_train, y_train, verbose = 2)
model_1.evaluate(x_test, y_test, verbose = 2)

1875/1875 - 4s - loss: 2.3026 - sparse_categorical_accuracy: 0.1000
313/313 - 1s - loss: 2.3026 - sparse_categorical_accuracy: 0.1000


[2.3026394844055176, 0.10000000149011612]

The accuracy is 0.10

In [9]:
model_1.summary()


Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
reshape (Reshape)            (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d (Conv2D)              (None, 28, 28, 32)        320       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 5, 5, 32)          0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 800)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 32)                25632     
_________________________________________________________________
dense_2 (Dense)              (None, 10)                330       
Total params: 26,282
Trainable params: 26,282
Non-trainable params: 0
__________________________________________________

- Add another four layers to the network: a convolution layer with 64 filters and a 3x3 max-pooling layer, followed by another convolution layer with 128 filters and a 3x3 max-pooling layer. Train the new network for 100 epochs and report the accuracy. <span style="color:red" float:right>[5 point]</span>

In [10]:
## your code goes here
model_2 = Sequential()
model_2.add(Reshape((28, 28, 1), input_shape=(28, 28)))

model_2.add(Conv2D(32, (3, 3), input_shape=(28, 28, 1), padding='same', activation='sigmoid'))
model_2.add(MaxPooling2D(pool_size=(3, 3)))

model_2.add(Conv2D(64, (3, 3), input_shape=(28, 28, 1), padding='same', activation='sigmoid'))
model_2.add(MaxPooling2D(pool_size=(3,3)))

model_2.add(Conv2D(128, (3, 3), input_shape=(28, 28, 1), padding='same', activation='sigmoid'))
model_2.add(MaxPooling2D(pool_size=(3, 3)))



model_2.add(Flatten())
model_2.add(Dense(32, activation='relu'))
model_2.add(Dense(10, activation = 'sigmoid'))


In [11]:

model_2.compile(
    optimizer = keras.optimizers.RMSprop(learning_rate = 1e-3),
    loss = keras.losses.SparseCategoricalCrossentropy(),
    metrics = [keras.metrics.SparseCategoricalAccuracy()],
)

In [12]:
model_2.fit(x_train, y_train, epochs = 100, verbose = 0)


In [13]:

model_2.evaluate(x_train, y_train, verbose = 2)
model_2.evaluate(x_test, y_test, verbose = 2)

1875/1875 - 6s - loss: 2.3026 - sparse_categorical_accuracy: 0.1000
313/313 - 2s - loss: 2.3026 - sparse_categorical_accuracy: 0.1000


[2.302640199661255, 0.10000000149011612]

The accuracy is 0.10

In [14]:
prediction = model_2.predict(x_test)

In [15]:
test_loss, test_acc = model_2.evaluate(x_test, y_test)

print('Test accuracy:', test_acc)

313/313 [==============================] - 1s 4ms/step - loss: 2.3026 - sparse_categorical_accuracy: 0.1000
Test accuracy: 0.10000000149011612


- How many parameters did we add to the model by adding the layers in the previous step? <span style="color:red" float:right>[5 point]</span>

# End of assignment

In [16]:
model_2.summary()


Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
reshape_1 (Reshape)          (None, 28, 28, 1)         0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 28, 28, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 9, 9, 32)          0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 9, 9, 64)          18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 3, 3, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 128)         73856     
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 1, 1, 128)        

##### The total of the model without the two last layers are 26,282
##### and the Total params of the lasr model are 97,130
##### So by adding the two layers, there will be 70,848 parameters.